### ISSUE TITLE
Error Reading in prot.accession2taxid.gz file

### ISSUE THREAD

**Issue by jnicholls01** [sherrillmix/taxonomizr#25]
hello, 

I have some protein accession numbers I'm trying to assign taxonomies too. Every time I try to prepare the database, either manually or in the all in one command, I keep getting this read error:

Error in readBin(inn, what = raw(0L), size = 1L, n = BFR.SIZE) : 
  error reading from the connection
In addition: Warning message:
In readBin(inn, what = raw(0L), size = 1L, n = BFR.SIZE) :
  invalid or incomplete compressed data

I have re-downloaded the prot.accession2taxid file multiple times, using the R commands and directly from the FTP website. The file always seems to be the right size (just over 6gb) so I'm quite stuck on what to try next...

Many Thanks,
Jack

**Comment by sherrillmix** [comment]
Could you post the results of `sessionInfo()` to start narrowing down what could be causing an issue?

Also since you have the file downloaded and convenient, could you run `zcat nucl_gb.accession2taxid.gz |tail` (might take a minute or so to process) or the equivalent and see what the end of the file looks like? It should be something like:

```
Z99997	Z99997.1	77599	3036826
Z99998	Z99998.1	77600	3036827
Z99999	Z99999.1	77601	3399750
```
without any gibberish or ending in the middle of a line.

**Comment by jnicholls01** [comment]
Thanks, there's something clearly very messed up with the file. The MD5 checksums also don't match... All the other files downloaded and can be read completely fine so I'm at a loss. I've downloaded the protein file many many times from R and the ftp site and with multiple internet connections. Any idea about what could be wrong?


> sessionInfo()
R version 4.0.3 (2020-10-10)
Platform: x86_64-w64-mingw32/x64 (64-bit)
Running under: Windows 10 x64 (build 19041)

Matrix products: default

locale:
[1] LC_COLLATE=English_United Kingdom.1252  LC_CTYPE=English_United Kingdom.1252   
[3] LC_MONETARY=English_United Kingdom.1252 LC_NUMERIC=C                           
[5] LC_TIME=English_United Kingdom.1252    

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] taxonomizr_0.5.3

loaded via a namespace (and not attached):
 [1] Rcpp_1.0.6        digest_0.6.27     R.methodsS3_1.8.1 DBI_1.1.1         RSQLite_2.2.2     rlang_0.4.10     
 [7] data.table_1.13.6 blob_1.2.1        R.oo_1.24.0       R.utils_2.10.1    vctrs_0.3.6       tools_4.0.3      
[13] bit64_4.0.5       tinytex_0.29      bit_4.0.4         xfun_0.20         compiler_4.0.3    pkgconfig_2.0.3  
[19] memoise_1.1.0    

jackn@LAPTOP-LCMRRD4N:/mnt/c/Users/jackn/Documents/rtemp$ zcat prot.accession2taxid.gz |tail

gzip: prot.accession2taxid.gz: invalid compressed data--format violated
85137146943441451A177A447447K8847K8113714744614145198444447447K844734174731744618694398444917914K8447341744744714186943441459179143447316887337A776
6447184B451984448947544279163341399948  43471194187917917719727631A15163784141468471144984489451714498743447K8116341154478198448947544917447885447414887716344487111731474471843464675145117B17A414163441677817881448478457348444918212414781161174844177519534433711791434474148871141345419841241138419453447544271A847314711474189714147433741344271A8474744541711883571445718147488044B45198444917914341494547B34497497534439418791791434475441114449149498114764734474377574947319647470477815349377574947387781788144494414787477447K88714414737371A84747446741897141719727631141377284434823545744978571444423443424747545744978581897141719727631441984123116544578374195171449816914941417484441474163163538748471449462944435117374191634731449462944484941417484441474477449174478447444
14844747884458
144649
547728574718178417716786943448117983437747341861635714444249116314444811A459414181778687636
442414737110
81914789714831981778687479834377474514147419418971411417494868774447311971397B754438663944717316836844811375834304778153457147433711789484714144444B44143447845744831478714738571444077134484144747449447873751815A1738571467714744434378348148447784141A447447K84414688114784448898574784195837741743737817544A4241486347314494684857B144744754153479179144485444474484857B144744178844554474784574994191634744177167869744541744517741944781484459631442491163143013714744479948      434774241789444737110
8187471B4799514444344741747785714745141474114141A4444733447774164198417444397B75443884398485187471B47994444948845A4781534571491437874484141A444417K88474774338813414765444414444364654457837414244498613541434A4778844554143781844476
442414748447784141K7514414114841344888983437747344241486347A17448737581738147147316836848449143781419841748455447478451789484714B4B
148413414844473444517741944185714444247864715811744444B44144479948      4185714847474711414444443448414499514444341K844744444418        7713444484434418447334477711744744717884574484495844B8214187471B479171471684494784195886231

**Comment by sherrillmix** [comment]
Just for posterity, this seems similar to the issue reported at the bottom of a previous issue #12. Although I never really got enough information to successfully debug that one. 

Here if I do `getAccession2taxid(types='prot')` I can replicate a similar error:

```
Error in (function (url, destfile, method, quiet = FALSE, mode = "w",  : 
  cannot open URL 'ftp://ftp.ncbi.nih.gov/pub/taxonomy/accession2taxid//prot.accession2taxid.gz'
In addition: Warning messages:
1: In (function (url, destfile, method, quiet = FALSE, mode = "w",  :
  downloaded length 4741857672 != reported length 6458643029
2: In (function (url, destfile, method, quiet = FALSE, mode = "w",  :
  URL 'ftp://ftp.ncbi.nih.gov/pub/taxonomy/accession2taxid//prot.accession2taxid.gz': Timeout of 60 seconds was reached
```

And it is replicable but with a different download length each time e.g. `downloaded length 5492033768 != reported length 6458643029` and `downloaded length 5524703328 != reported length 6458643029`.

If I try outside R with `wget ftp://ftp.ncbi.nih.gov/pub/taxonomy/accession2taxid//prot.accession2taxid.gz*` then I get a successful download

```
2021-02-05 09:11:42 (101 MB/s) - ‘prot.accession2taxid.gz’ saved [6458643029]
```

where the md5sum matches what is expected:

```
>md5sum prot.accession2taxid.gz
a40aa440ad68189fc635683ab2a98e12  prot.accession2taxid.gz
>cat prot.accession2taxid.gz.md5 
a40aa440ad68189fc635683ab2a98e12  prot.accession2taxid.gz
```

That plus the "Timeout of 60 seconds was reached" warning makes me suspicious that the `download.file` function in R may have a "helpful" feature timing out long downloads which would not be optimal when we're working with these huge files. Checking now.

**Comment by sherrillmix** [comment]
Yeah it is indeed a "helpful" R feature. I'll fix in the package. If you were in a hurry you could probably call `options(timeout=10*60*60)` to set the timeout to 10 hours then proceed with `taxonomizr` as normal.

 Although if you were also having problems downloading outside R then there may be additional issues.

**Comment by jnicholls01** [comment]
Ah that is very 'helpful'...

I guess it was just timing out on my too slow internet. It appears to have downloaded correctly this time so thanks very much for your help!

**Comment by sherrillmix** [comment]
There's a `timeout` option included now in v0.6.0. It defaults to 10 hours so hopefully that will forestall any similar issues in the future.
